### PR TITLE
feat(getGuildBans): support pagination

### PR DESCRIPTION
**Pull request by HeadTriXz**
Adds support for guild bans pagination: https://discord.com/developers/docs/resources/guild#get-guild-bans

### CODE REVIEW

File: lib/Client.js [hunk 1]
```diff
@@ -2332,15 +2332,41 @@ class Client extends EventEmitter {
...
+    * @returns {Promise<Array<Object>>} Resolves with an array of { reason: String, user: User }
+    */
+    async getGuildBans(guildID, options = {}) {
+        const bans = await this.requestHandler.request("GET", Endpoints.GUILD_BANS(guildID), true, {
+            after: options.after,
+            before: options.before,
+            limit: Math.min(options.limit, 1000) 
         });
+        
+        bans.forEach((ban) => {
```
Comment (ghost) [inline lib/Client.js]: forEach isn't synchronous

Comment (HeadTriXz) [reply lib/Client.js]: Discussed in Discord.

**Review by ghost — DISMISSED** [review]

File: lib/Client.js (hunk 1 above)
Comment (bsian03) [inline lib/Client.js]: Change this to `map`

File: lib/Client.js [hunk 2]
```diff
@@ -2332,15 +2332,41 @@ class Client extends EventEmitter {
...
-    getGuildBans(guildID) {
-        return this.requestHandler.request("GET", Endpoints.GUILD_BANS(guildID), true).then((bans) => {
-            bans.forEach((ban) => {
-                ban.user = new User(ban.user, this);
-            });
-            return bans;
+    * @arg {Object} [options] Options for the request
+    * @arg {String} [options.after] Only get users after given user ID
+    * @arg {String} [options.before] Only get users before given user ID
+    * @arg {Number} [options.limit] The maximum number of users to return
```
Comment (bsian03) [inline lib/Client.js]: Add the default limit here

File: lib/Client.js [hunk 3]
```diff
@@ -2332,15 +2332,41 @@ class Client extends EventEmitter {
...
+            limit: Math.min(options.limit, 1000) 
         });
+        
+        bans.forEach((ban) => {
+            ban.user = this.users.update(ban.user, this);
+        });
+        
+        if(options.limit && options.limit > 1000 && bans.length >= 1000) {
+            if(options.before) {
+                options.after = undefined;
```
Comment (bsian03) [inline lib/Client.js]: May cause unintended side effects if a user supplies both options.before and options.after

**Review by bsian03 — DISMISSED** [review]

File: lib/Client.js (hunk 3 above)
Comment (HeadTriXz) [reply lib/Client.js]: Discord API docs state that they will use `before` if both are supplied.

File: lib/Client.js (hunk 2 above)
Comment (HeadTriXz) [reply lib/Client.js]: What's the default limit?

File: lib/Client.js (hunk 1 above)
Comment (DonovanDMC) [reply lib/Client.js]: Why should this be a map? We aren't returning a value

Comment (HeadTriXz) [reply lib/Client.js]: `.map` will return an array with new objects. `.forEach` was used before and only updates the value of the property. An other option would be to use a for loop.

File: lib/Client.js (hunk 2 above)
Comment (DonovanDMC) [reply lib/Client.js]: https://discord.com/developers/docs/resources/guild#get-guild-bans

1000

File: lib/Client.js (hunk 1 above)
Comment (bsian03) [reply lib/Client.js]: Should also say move this logic to the return itself with the `.map`

Comment (HeadTriXz) [reply lib/Client.js]: Would mess up the recursion.

Comment (bsian03) [reply lib/Client.js]: Resolved within discord. turns out i thought similar circumstance logic was different. whoops

**Review by DonovanDMC — DISMISSED** [review]

File: lib/Client.js [hunk 4]
```diff
@@ -2332,15 +2332,39 @@ class Client extends EventEmitter {
...
+        if(options.limit && options.limit > 1000 && bans.length >= 1000) {
+            const page = await this.getGuildBans(guildID, {
+                after: options.before ? undefined : bans[bans.length - 1].user.id,
+                before: options.before ? bans[0].user.id : undefined,
+                limit: options.limit - bans.length
+            });
+
+            if(options.before) {
+                page.push(...bans);
+                bans = page;
```
Comment (HeadTriXz) [inline lib/Client.js]: `Array#unshift` exists.

Comment (DonovanDMC) [reply lib/Client.js]: we literally.. already talked about that

Comment (HeadTriXz) [reply lib/Client.js]: I have great wifi in the train 🚂

**Review by abalabahaha — APPROVED** [review]

**Pull request merged**